File: android/Gutenberg/src/test/java/org/wordpress/gutenberg/http/ChunkedMultipartTests.kt
package org.wordpress.gutenberg.http

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Test
import java.io.File

/**
 * Tests for the chunked (file-backed) multipart parsing path.
 *
 * The in-memory path is tested extensively in [FixtureTests] and the shared
 * JSON fixture files. These tests verify the chunked scanner that runs when
 * the body is backed by a file on disk.
 */
class ChunkedMultipartTests {

    // MARK: - Basic Parsing

    @Test
    fun `single text field parsed from file-backed body`() {
        val (file, request) = makeFileBackedRequest(
            fields = listOf(Field("title", value = "My Blog Post".toByteArray())),
            boundary = "AaB03x"
        )
        file.deleteOnExit()

        val parts = request.multipartParts()

        assertEquals(1, parts.size)
        assertEquals("title", parts[0].name)
        assertEquals(null, parts[0].filename)
        assertEquals("text/plain", parts[0].contentType)
        assertArrayEquals("My Blog Post".toByteArray(), parts[0].body.readBytes())
    }

    @Test
    fun `multiple parts parsed from file-backed body`() {
        val (file, request) = makeFileBackedRequest(
            fields = listOf(
                Field("title", value = "Hello".toByteArray()),
                Field("file", filename = "photo.jpg", contentType = "image/jpeg", value = "jpeg-data".toByteArray()),
                Field("caption", value = "A photo".toByteArray()),
            ),
            boundary = "WebKitBoundary123"
        )
        file.deleteOnExit()

        val parts = request.multipartParts()

        assertEquals(3, parts.size)
        assertEquals("title", parts[0].name)
        assertArrayEquals("Hello".toByteArray(), parts[0].body.readBytes())
        assertEquals("file", parts[1].name)
        assertEquals("photo.jpg", parts[1].filename)
        assertEquals("image/jpeg", parts[1].contentType)
        assertArrayEquals("jpeg-data".toByteArray(), parts[1].body.readBytes())
        assertEquals("caption", parts[2].name)
        assertArrayEquals("A photo".toByteArray(), parts[2].body.readBytes())
    }

    @Test
    fun `empty part body parsed correctly`() {
        val (file, request) = makeFileBackedRequest(
            fields = listOf(Field("empty", value = ByteArray(0))),
            boundary = "AaB03x"
        )
        file.deleteOnExit()

        val parts = request.multipartParts()

        assertEquals(1, parts.size)
        assertEquals("empty", parts[0].name)
        assertArrayEquals(ByteArray(0), parts[0].body.readBytes())
    }

    @Test
    fun `binary data preserved through file-backed parsing`() {
        val binaryContent = ByteArray(128) { 0x00 } +
            ByteArray(256) { it.toByte() } +
            ByteArray(128) { 0xFF.toByte() }

        val (file, request) = makeFileBackedRequest(
            fields = listOf(Field("file", filename = "binary.bin", contentType = "application/octet-stream", value = binaryContent)),
            boundary = "BinaryBoundary99"
        )
        file.deleteOnExit()

        val parts = request.multipartParts()

        assertEquals(1, parts.size)
        assertEquals("binary.bin", parts[0].filename)
        assertArrayEquals(binaryContent, parts[0].body.readBytes())
    }

    @Test
    fun `preamble before first boundary is ignored`() {
        val boundary = "AaB03x"
        val body = "This is the preamble. It should be ignored.\r\n" +
            "--$boundary\r\nContent-Disposition: form-data; name=\"field\"\r\n\r\nvalue\r\n--$boundary--\r\n"

        val (file, request) = makeFileBackedRequestFromRawBody(body, boundary)
        file.deleteOnExit()

        val parts = request.multipartParts()

        assertEquals(1, parts.size)
        assertEquals("field", parts[0].name)
        assertArrayEquals("value".toByteArray(), parts[0].body.readBytes())
    }

    @Test
    fun `transport padding after boundary is skipped`() {
        val boundary = "AaB03x"
        val body = "--$boundary  \t \r\nContent-Disposition: form-data; name=\"field\"\r\n\r\nvalue\r\n--$boundary--\r\n"

        val (file, request) = makeFileBackedRequestFromRawBody(body, boundary)
        file.deleteOnExit()

        val parts = request.multipartParts()

        assertEquals(1, parts.size)
        assertEquals("field", parts[0].name)
        assertArrayEquals("value".toByteArray(), parts[0].body.readBytes())
    }

    // MARK: - Error Cases

    @Test(expected = MultipartParseException::class)
    fun `close-delimiter-only body throws malformedBody`() {
        val boundary = "AaB03x"
        val body = "--$boundary--\r\n"

        val (file, request) = makeFileBackedRequestFromRawBody(body, boundary)
        file.deleteOnExit()

        request.multipartParts()
    }

    @Test(expected = MultipartParseException::class)
    fun `missing close delimiter throws malformedBody`() {
        val boundary = "AaB03x"
        val body = "--$boundary\r\nContent-Disposition: form-data; name=\"field\"\r\n\r\nvalue"

        val (file, request) = makeFileBackedRequestFromRawBody(body, boundary)
        file.deleteOnExit()

        request.multipartParts()
    }

    // MARK: - Chunk Boundary Edge Cases

    @Test
    fun `boundary split across chunk boundary is found correctly`() {
        val boundary = "AaB03x"
        val delimiter = "--$boundary" // 10 bytes

        // We want the second delimiter to start 5 bytes before the 65536 chunk boundary.
        val splitPoint = 65_536 - 5

        val headerBytes = "--$boundary\r\nContent-Disposition: form-data; name=\"pad\"\r\n\r\n".toByteArray()
        val headerOverhead = headerBytes.size
        val crlfBeforeDelimiter = 2
        val paddingLength = splitPoint - headerOverhead - crlfBeforeDelimiter

        val padding = ByteArray(paddingLength) { 'A'.code.toByte() }

        val (file, request) = makeFileBackedRequest(
            fields = listOf(
                Field("pad", value = padding),
                Field("after", value = "found-it".toByteArray()),
            ),
            boundary = boundary
        )
        file.deleteOnExit()

        // Verify the delimiter actually straddles the chunk boundary.
        val fileData = file.readBytes()
        val delimBytes = delimiter.toByteArray()
        val delimStart = ReadOnlyBytes(fileData).indexOf(delimBytes, headerOverhead)
        assertEquals("Delimiter should start at $splitPoint", splitPoint, delimStart)

        val parts = request.multipartParts()

        assertEquals(2, parts.size)
        assertEquals("pad", parts[0].name)
        assertEquals(paddingLength.toLong(), parts[0].body.size)
        assertEquals("after", parts[1].name)
        assertArrayEquals("found-it".toByteArray(), parts[1].body.readBytes())
    }

    @Test
    fun `large body spanning multiple chunks parses correctly`() {
        val largeContent = ByteArray(200_000) { 'X'.code.toByte() }

        val (file, request) = makeFileBackedRequest(
            fields = listOf(
                Field("large", filename = "big.bin", contentType = "application/octet-stream", value = largeContent),
                Field("meta", value = "description".toByteArray()),
            ),
            boundary = "LargeBoundary42"
        )
        file.deleteOnExit()

        val parts = request.multipartParts()

        assertEquals(2, parts.size)
        assertEquals("large", parts[0].name)
        assertEquals(largeContent.size.toLong(), parts[0].body.size)
        assertArrayEquals(largeContent, parts[0].body.readBytes())
        assertEquals("meta", parts[1].name)
        assertArrayEquals("description".toByteArray(), parts[1].body.readBytes())
    }

    // MARK: - fileSlice Source

    @Test
    fun `file-backed body with non-zero offset (fileSlice) parses correctly`() {
        val boundary = "AaB03x"
        val multipartBody = "--$boundary\r\nContent-Disposition: form-data; name=\"field\"\r\n\r\nvalue\r\n--$boundary--\r\n"
        val multipartData = multipartBody.toByteArray()

        val garbagePrefix = ByteArray(500) { 'Z'.code.toByte() }
        val file = File.createTempFile("slice-test-", null)
        file.deleteOnExit()
        file.writeBytes(garbagePrefix + multipartData)

        val body = RequestBody.FileBacked(
            file = file,
            fileOffset = garbagePrefix.size.toLong(),
            size = multipartData.size.toLong()
        )
        val request = ParsedHTTPRequest(
            method = "POST",
            target = "/upload",
            httpVersion = "HTTP/1.1",
            headers = mapOf("Content-Type" to "multipart/form-data; boundary=$boundary", "Host" to "localhost"),
            body = body,
            isComplete = true
        )

        val parts = request.multipartParts()

        assertEquals(1, parts.size)
        assertEquals("field", parts[0].name)
        assertArrayEquals("value".toByteArray(), parts[0].body.readBytes())
    }

    // MARK: - Part Count Limit

    @Test(expected = MultipartParseException::class)
    fun `rejects multipart body with more than 100 parts (in-memory)`() {
        val fields = (0 until 101).map { Field("field$it", value = "val$it".toByteArray()) }
        val bodyData = buildMultipartBody(fields, "AaB03x")
        val body = RequestBody.InMemory(bodyData)
        val request = ParsedHTTPRequest(
            method = "POST",
            target = "/upload",
            httpVersion = "HTTP/1.1",
            headers = mapOf("Content-Type" to "multipart/form-data; boundary=AaB03x", "Host" to "localhost"),
            body = body,
            isComplete = true
        )
        request.multipartParts()
    }

    @Test
    fun `accepts multipart body with exactly 100 parts (in-memory)`() {
        val fields = (0 until 100).map { Field("field$it", value = "val$it".toByteArray()) }
        val bodyData = buildMultipartBody(fields, "AaB03x")
        val body = RequestBody.InMemory(bodyData)
        val request = ParsedHTTPRequest(
            method = "POST",
            target = "/upload",
            httpVersion = "HTTP/1.1",
            headers = mapOf("Content-Type" to "multipart/form-data; boundary=AaB03x", "Host" to "localhost"),
            body = body,
            isComplete = true
        )
        val parts = request.multipartParts()
        assertEquals(100, parts.size)
    }

    @Test(expected = MultipartParseException::class)
    fun `rejects multipart body with more than 100 parts (file-backed)`() {
        val fields = (0 until 101).map { Field("field$it", value = "val$it".toByteArray()) }
        val (file, request) = makeFileBackedRequest(fields, "AaB03x")
        file.deleteOnExit()
        request.multipartParts()
    }

    @Test
    fun `accepts multipart body with exactly 100 parts (file-backed)`() {
        val fields = (0 until 100).map { Field("field$it", value = "val$it".toByteArray()) }
        val (file, request) = makeFileBackedRequest(fields, "AaB03x")
        file.deleteOnExit()
        val parts = request.multipartParts()
        assertEquals(100, parts.size)
    }

    // MARK: - Helpers

    private data class Field(
        val name: String,
        val filename: String? = null,
        val contentType: String? = null,
        val value: ByteArray
    )

    private fun makeFileBackedRequest(
        fields: List<Field>,
        boundary: String
    ): Pair<File, ParsedHTTPRequest> {
        val body = buildMultipartBody(fields, boundary)

        val file = File.createTempFile("multipart-test-", null)
        file.writeBytes(body)

        val requestBody = RequestBody.FileBacked(
            file = file,
            fileOffset = 0,
            size = body.size.toLong()
        )
        val request = ParsedHTTPRequest(
            method = "POST",
            target = "/upload",
            httpVersion = "HTTP/1.1",
            headers = mapOf("Content-Type" to "multipart/form-data; boundary=$boundary", "Host" to "localhost"),
            body = requestBody,
            isComplete = true
        )
        return file to request
    }

    private fun makeFileBackedRequestFromRawBody(
        body: String,
        boundary: String
    ): Pair<File, ParsedHTTPRequest> {
        val bodyData = body.toByteArray()
        val file = File.createTempFile("multipart-test-", null)
        file.writeBytes(bodyData)

        val requestBody = RequestBody.FileBacked(
            file = file,
            fileOffset = 0,
            size = bodyData.size.toLong()
        )
        val request = ParsedHTTPRequest(
            method = "POST",
            target = "/upload",
            httpVersion = "HTTP/1.1",
            headers = mapOf("Content-Type" to "multipart/form-data; boundary=$boundary", "Host" to "localhost"),
            body = requestBody,
            isComplete = true
        )
        return file to request
    }

    private fun buildMultipartBody(fields: List<Field>, boundary: String): ByteArray {
        val out = java.io.ByteArrayOutputStream()
        for (field in fields) {
            out.write("--$boundary\r\n".toByteArray())
            var disposition = "Content-Disposition: form-data; name=\"${field.name}\""
            if (field.filename != null) {
                disposition += "; filename=\"${field.filename}\""
            }
            out.write("$disposition\r\n".toByteArray())
            if (field.contentType != null) {
                out.write("Content-Type: ${field.contentType}\r\n".toByteArray())
            }
            out.write("\r\n".toByteArray())
            out.write(field.value)
            out.write("\r\n".toByteArray())
        }
        out.write("--$boundary--\r\n".toByteArray())
        return out.toByteArray()
    }
}

File: ios/Tests/GutenbergKitHTTPTests/ChunkedMultipartTests.swift
import Foundation
import Testing
@testable import GutenbergKitHTTP

/// Tests for the chunked (file-backed) multipart parsing path.
///
/// The in-memory path is tested extensively in ``RFC7578ConformanceTests`` and
/// the shared fixture tests. These tests verify the chunked scanner that runs
/// when the body is backed by a file on disk.
@Suite("Chunked Multipart Parsing")
struct ChunkedMultipartTests {

    // MARK: - Basic Parsing

    @Test("single text field parsed from file-backed body")
    func singleTextField() throws {
        let (url, request) = try makeFileBackedRequest(
            fields: [("title", nil, nil, Data("My Blog Post".utf8))],
            boundary: "AaB03x"
        )
        defer { try? FileManager.default.removeItem(at: url) }

        let parts = try request.multipartParts()

        #expect(parts.count == 1)
        #expect(parts[0].name == "title")
        #expect(parts[0].filename == nil)
        #expect(parts[0].contentType == "text/plain")
        #expect(try readAll(parts[0].body) == Data("My Blog Post".utf8))
    }

    @Test("multiple parts parsed from file-backed body")
    func multipleParts() throws {
        let (url, request) = try makeFileBackedRequest(
            fields: [
                ("title", nil, nil, Data("Hello".utf8)),
                ("file", "photo.jpg", "image/jpeg", Data("jpeg-data".utf8)),
                ("caption", nil, nil, Data("A photo".utf8)),
            ],
            boundary: "WebKitBoundary123"
        )
        defer { try? FileManager.default.removeItem(at: url) }

        let parts = try request.multipartParts()

        #expect(parts.count == 3)
        #expect(parts[0].name == "title")
        #expect(try readAll(parts[0].body) == Data("Hello".utf8))
        #expect(parts[1].name == "file")
        #expect(parts[1].filename == "photo.jpg")
        #expect(parts[1].contentType == "image/jpeg")
        #expect(try readAll(parts[1].body) == Data("jpeg-data".utf8))
        #expect(parts[2].name == "caption")
        #expect(try readAll(parts[2].body) == Data("A photo".utf8))
    }

    @Test("empty part body parsed correctly")
    func emptyPartBody() throws {
        let (url, request) = try makeFileBackedRequest(
            fields: [("empty", nil, nil, Data())],
            boundary: "AaB03x"
        )
        defer { try? FileManager.default.removeItem(at: url) }

        let parts = try request.multipartParts()

        #expect(parts.count == 1)
        #expect(parts[0].name == "empty")
        #expect(try readAll(parts[0].body) == Data())
    }

    @Test("binary data preserved through file-backed parsing")
    func binaryData() throws {
        // Include bytes that would be problematic if treated as text: NUL, 0xFF, CRLF sequences.
        var binaryContent = Data(repeating: 0x00, count: 128)
        binaryContent.append(contentsOf: (0...255).map { UInt8($0) })
        binaryContent.append(Data(repeating: 0xFF, count: 128))

        let (url, request) = try makeFileBackedRequest(
            fields: [("file", "binary.bin", "application/octet-stream", binaryContent)],
            boundary: "BinaryBoundary99"
        )
        defer { try? FileManager.default.removeItem(at: url) }

        let parts = try request.multipartParts()

        #expect(parts.count == 1)
        #expect(parts[0].filename == "binary.bin")
        #expect(try readAll(parts[0].body) == binaryContent)
    }

    @Test("preamble before first boundary is ignored")
    func preambleIgnored() throws {
        let boundary = "AaB03x"
        let preamble = "This is the preamble. It should be ignored.\r\n"
        let body = "\(preamble)--\(boundary)\r\nContent-Disposition: form-data; name=\"field\"\r\n\r\nvalue\r\n--\(boundary)--\r\n"

        let (url, request) = try makeFileBackedRequestFromRawBody(body: body, boundary: boundary)
        defer { try? FileManager.default.removeItem(at: url) }

        let parts = try request.multipartParts()

        #expect(parts.count == 1)
        #expect(parts[0].name == "field")
        #expect(try readAll(parts[0].body) == Data("value".utf8))
    }

    @Test("transport padding after boundary is skipped")
    func transportPadding() throws {
        let boundary = "AaB03x"
        // Add spaces and tabs after the boundary delimiter
        let body = "--\(boundary)  \t \r\nContent-Disposition: form-data; name=\"field\"\r\n\r\nvalue\r\n--\(boundary)--\r\n"

        let (url, request) = try makeFileBackedRequestFromRawBody(body: body, boundary: boundary)
        defer { try? FileManager.default.removeItem(at: url) }

        let parts = try request.multipartParts()

        #expect(parts.count == 1)
        #expect(parts[0].name == "field")
        #expect(try readAll(parts[0].body) == Data("value".utf8))
    }

    // MARK: - Error Cases

    @Test("close-delimiter-only body throws malformedBody")
    func closeDelimiterOnly() throws {
        let boundary = "AaB03x"
        let body = "--\(boundary)--\r\n"

        let (url, request) = try makeFileBackedRequestFromRawBody(body: body, boundary: boundary)
        defer { try? FileManager.default.removeItem(at: url) }

        #expect(throws: MultipartParseError.malformedBody) {
            try request.multipartParts()
        }
    }

    @Test("missing close delimiter throws malformedBody")
    func missingCloseDelimiter() throws {
        let boundary = "AaB03x"
        let body = "--\(boundary)\r\nContent-Disposition: form-data; name=\"field\"\r\n\r\nvalue"

        let (url, request) = try makeFileBackedRequestFromRawBody(body: body, boundary: boundary)
        defer { try? FileManager.default.removeItem(at: url) }

        #expect(throws: MultipartParseError.malformedBody) {
            try request.multipartParts()
        }
    }

    // MARK: - Chunk Boundary Edge Cases

    @Test("boundary split across chunk boundary is found correctly")
    func boundarySplitAcrossChunk() throws {
        let boundary = "AaB03x"
        let delimiter = "--\(boundary)"  // 10 bytes

        // We want the second delimiter to start 5 bytes before the 65536 chunk boundary,
        // so it straddles the boundary: 5 bytes in chunk 1, 5 bytes in chunk 2.
        let splitPoint = 65_536 - 5

        // Calculate the header overhead for the first part:
        //   "--AaB03x\r\n" = 12 bytes
        //   "Content-Disposition: form-data; name=\"pad\"\r\n" = 45 bytes
        //   "\r\n" = 2 bytes  (header/body separator)
        // Total: 59 bytes
        // After the body: "\r\n" = 2 bytes (CRLF before next delimiter)
        // So: padding_length = splitPoint - 59 - 2 = splitPoint - 61
        let headerOverhead = Data("--\(boundary)\r\nContent-Disposition: form-data; name=\"pad\"\r\n\r\n".utf8).count
        let crlfBeforeDelimiter = 2
        let paddingLength = splitPoint - headerOverhead - crlfBeforeDelimiter

        let padding = Data(repeating: UInt8(ascii: "A"), count: paddingLength)

        let (url, request) = try makeFileBackedRequest(
            fields: [
                ("pad", nil, nil, padding),
                ("after", nil, nil, Data("found-it".utf8)),
            ],
            boundary: boundary
        )
        defer { try? FileManager.default.removeItem(at: url) }

        // Verify the delimiter actually straddles the chunk boundary.
        let fileData = try Data(contentsOf: url)
        let delimData = Data(delimiter.utf8)
        if let range = fileData.range(of: delimData, in: (fileData.startIndex + headerOverhead)..<fileData.endIndex) {
            let delimStart = fileData.distance(from: fileData.startIndex, to: range.lowerBound)
            #expect(delimStart == splitPoint, "Delimiter should start at \(splitPoint), got \(delimStart)")
        }

        let parts = try request.multipartParts()

        #expect(parts.count == 2)
        #expect(parts[0].name == "pad")
        #expect(parts[0].body.count == paddingLength)
        #expect(parts[1].name == "after")
        #expect(try readAll(parts[1].body) == Data("found-it".utf8))
    }

    @Test("large body spanning multiple chunks parses correctly")
    func largeBodyMultipleChunks() throws {
        // Create a body larger than 2 chunks (> 128 KB).
        let largeContent = Data(repeating: UInt8(ascii: "X"), count: 200_000)

        let (url, request) = try makeFileBackedRequest(
            fields: [
                ("large", "big.bin", "application/octet-stream", largeContent),
                ("meta", nil, nil, Data("description".utf8)),
            ],
            boundary: "LargeBoundary42"
        )
        defer { try? FileManager.default.removeItem(at: url) }

        let parts = try request.multipartParts()

        #expect(parts.count == 2)
        #expect(parts[0].name == "large")
        #expect(parts[0].body.count == largeContent.count)
        #expect(try readAll(parts[0].body) == largeContent)
        #expect(parts[1].name == "meta")
        #expect(try readAll(parts[1].body) == Data("description".utf8))
    }

    // MARK: - fileSlice Source

    @Test("file-backed body with non-zero offset (fileSlice) parses correctly")
    func fileSliceSource() throws {
        let boundary = "AaB03x"
        let multipartBody = "--\(boundary)\r\nContent-Disposition: form-data; name=\"field\"\r\n\r\nvalue\r\n--\(boundary)--\r\n"
        let multipartData = Data(multipartBody.utf8)

        // Write garbage prefix + multipart body to the file.
        let garbagePrefix = Data(repeating: UInt8(ascii: "Z"), count: 500)
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("slice-test-\(UUID().uuidString)")
        try (garbagePrefix + multipartData).write(to: url)
        defer { try? FileManager.default.removeItem(at: url) }

        // Create a fileSlice body that starts after the garbage prefix.
        let body = RequestBody(fileURL: url, offset: UInt64(garbagePrefix.count), length: multipartData.count)
        let request = ParsedHTTPRequest.complete(
            method: "POST",
            target: "/upload",
            httpVersion: "HTTP/1.1",
            headers: ["Content-Type": "multipart/form-data; boundary=\(boundary)", "Host": "localhost"],
            body: body
        )

        let parts = try request.multipartParts()

        #expect(parts.count == 1)
        #expect(parts[0].name == "field")
        #expect(try readAll(parts[0].body) == Data("value".utf8))
    }

    // MARK: - Helpers

    /// Builds a multipart body from field descriptors, writes it to a temp file,
    /// and returns a `ParsedHTTPRequest` with a file-backed body.
    private func makeFileBackedRequest(
        fields: [(name: String, filename: String?, contentType: String?, value: Data)],
        boundary: String
    ) throws -> (URL, ParsedHTTPRequest) {
        var body = Data()
        for field in fields {
            body.append(Data("--\(boundary)\r\n".utf8))
            var disposition = "Content-Disposition: form-data; name=\"\(field.name)\""
            if let filename = field.filename {
                disposition += "; filename=\"\(filename)\""
            }
            body.append(Data("\(disposition)\r\n".utf8))
            if let ct = field.contentType {
                body.append(Data("Content-Type: \(ct)\r\n".utf8))
            }
            body.append(Data("\r\n".utf8))
            body.append(field.value)
            body.append(Data("\r\n".utf8))
        }
        body.append(Data("--\(boundary)--\r\n".utf8))

        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("multipart-test-\(UUID().uuidString)")
        try body.write(to: url)

        let requestBody = RequestBody(fileURL: url)
        let request = ParsedHTTPRequest.complete(
            method: "POST",
            target: "/upload",
            httpVersion: "HTTP/1.1",
            headers: ["Content-Type": "multipart/form-data; boundary=\(boundary)", "Host": "localhost"],
            body: requestBody
        )

        return (url, request)
    }

    /// Writes a raw multipart body string to a temp file and returns a file-backed request.
    private func makeFileBackedRequestFromRawBody(
        body: String,
        boundary: String
    ) throws -> (URL, ParsedHTTPRequest) {
        let bodyData = Data(body.utf8)
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("multipart-test-\(UUID().uuidString)")
        try bodyData.write(to: url)

        let requestBody = RequestBody(fileURL: url)
        let request = ParsedHTTPRequest.complete(
            method: "POST",
            target: "/upload",
            httpVersion: "HTTP/1.1",
            headers: ["Content-Type": "multipart/form-data; boundary=\(boundary)", "Host": "localhost"],
            body: requestBody
        )

        return (url, request)
    }

    private func readAll(_ body: RequestBody) throws -> Data {
        let stream = try body.makeInputStream()
        stream.open()
        defer { stream.close() }

        var data = Data()
        let bufferSize = 1024
        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
        defer { buffer.deallocate() }

        // Use read() directly instead of hasBytesAvailable to avoid race
        // conditions with bound stream pairs, where data from the writer
        // thread may not have arrived yet when hasBytesAvailable is checked.
        while true {
            let bytesRead = stream.read(buffer, maxLength: bufferSize)
            if bytesRead <= 0 { break }
            data.append(buffer, count: bytesRead)
        }

        return data
    }
}
